metagraph/src/annotation/binary_matrix/multi_brwt/clustering.cpp

```diff
@@ -225,9 +259,63 @@ Partition greedy_matching(const std::vector<sdsl::bit_vector> &columns,
     return partition;
 }
 
-LinkageMatrix
-agglomerative_greedy_linkage(std::vector<sdsl::bit_vector>&& columns,
-                             size_t num_threads) {
+void merge(const sdsl::bit_vector &first, sdsl::bit_vector *second) {
+    assert(second);
+    assert(first.size() == second->size());
+    *second |= first;
+}
+
+void merge(const SparseColumn &first, SparseColumn *second) {
+    assert(second);
+
+    const auto &[size_first, col_first] = first;
+    const auto &[size_second, col_second] = *second;
+
+    SparseColumn merged;
+    auto &[size_merged, col_merged] = merged;
+
+    size_merged = std::min(size_first, size_second);
+
+    col_merged.reserve(col_first.size() + col_second.size());
+
+    auto it_first = col_first.begin();
+    auto it_second = col_second.begin();
+
+    while (it_first != col_first.end() && it_second != col_second.end()) {
+        if (*it_first < *it_second) {
+            col_merged.push_back(*it_first);
+            ++it_first;
+        } else if (*it_first > *it_second) {
+            col_merged.push_back(*it_second);
+            ++it_second;
+        } else {
+            col_merged.push_back(*it_first);
+            ++it_first;
+            ++it_second;
+        }
+    }
+    while (it_first != col_first.end() && *it_first < size_merged) {
+        col_merged.push_back(*it_first);
+        ++it_first;
+    }
+    while (it_second != col_second.end() && *it_second < size_merged) {
+        col_merged.push_back(*it_second);
+        ++it_second;
+    }
```

Comment:
why not just use `std::merge`?
```suggestion
    std::merge(col_first.begin(), col_first.end(),
               col_second.begin(), col_second.end(),
               std::back_inserter(col_merged));
```

Reply:
std::merge will have duplications

Reply:
ah, I looked into it again and I think `std::set_union` should work then

Reply:
Not exactly. I merge until the minimum size of these two.

Here we merge until the min(size_1, size_2).

I can make it work with std::set_union, but then I'd need to do a scan to find the new `end` first.

Is that worth it? I think the implementation of std::set_union is the same as here, anyway. No?